src/Mvc/Mvc.Core/test/Infrastructure/AsyncEnumerableReaderTest.cs

```diff
@@ -5,57 +5,168 @@
 using System.Collections.Generic;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.Extensions.Options;
 using Xunit;
 
 namespace Microsoft.AspNetCore.Mvc.Infrastructure
 {
     public class AsyncEnumerableReaderTest
     {
+        [Theory]
+        [InlineData(typeof(Range))]
+        [InlineData(typeof(IEnumerable<string>))]
+        [InlineData(typeof(List<int>))]
+        public void TryGetReader_ReturnsFalse_IfTypeIsNotIAsyncEnumerable(Type type)
+        {
+            // Arrange
+            var options = new MvcOptions();
+            var readerFactory = new AsyncEnumerableReader(options);
+            var asyncEnumerable = TestEnumerable();
+
+            // Act
+            var result = readerFactory.TryGetReader(type, out var reader);
+
+            // Assert
+            Assert.False(result);
+        }
+
         [Fact]
-        public async Task ReadAsync_ReadsIAsyncEnumerable()
+        public async Task TryGetReader_ReturnsReaderForIAsyncEnumerable()
         {
             // Arrange
             var options = new MvcOptions();
-            var reader = new AsyncEnumerableReader(options);
+            var readerFactory = new AsyncEnumerableReader(options);
+            var asyncEnumerable = TestEnumerable();
 
             // Act
-            var result = await reader.ReadAsync(TestEnumerable());
+            var result = readerFactory.TryGetReader(asyncEnumerable.GetType(), out var reader);
 
             // Assert
-            var collection = Assert.IsAssignableFrom<ICollection<string>>(result);
+            Assert.True(result);
+            var readCollection = await reader(asyncEnumerable);
+            var collection = Assert.IsAssignableFrom<ICollection<string>>(readCollection);
             Assert.Equal(new[] { "0", "1", "2", }, collection);
         }
 
         [Fact]
-        public async Task ReadAsync_ReadsIAsyncEnumerable_ImplementingMultipleAsyncEnumerableInterfaces()
+        public async Task TryGetReader_ReturnsReaderForIAsyncEnumerableOfValueType()
+        {
+            // Arrange
+            var options = new MvcOptions();
+            var readerFactory = new AsyncEnumerableReader(options);
+            var asyncEnumerable = PrimitiveEnumerable();
+
+            // Act
+            var result = readerFactory.TryGetReader(asyncEnumerable.GetType(), out var reader);
+
+            // Assert
+            Assert.True(result);
+            var readCollection = await reader(asyncEnumerable);
+            var collection = Assert.IsAssignableFrom<ICollection<int>>(readCollection);
+            Assert.Equal(new[] { 0, 1, 2, }, collection);
+        }
+
+        [Fact]
+        public void TryGetReader_ReturnsCachedDelegate()
```

Comment:
How does this behave when a type implements multiple IAsyncEnumerables<>

Reply:
We cache this based on the passed in type, so it just works. I added a test for good measure.